### PR TITLE
srml-contract: introduce ext_caller external function

### PR DESCRIPTION
**Pull request by pepyakin**
This PR introduces a new function called `ext_caller` to the contract runtime. This function is useful, not invasive and not controversial (in contract to the function for querying an origin of a transaction).

TODO: 

- [x] Update COMPLEXITY.md (and before that find where it's gone)

cc @Robbepop

### CODE REVIEW

File: srml/contract/COMPLEXITY.md [hunk 1]
```diff
@@ -275,6 +275,12 @@ This function receives a `data` buffer as an argument. Execution of the function
 
 **complexity**: The complexity of this function is proportional to the size of the `data` buffer.
 
+## ext_caller
+
+This function serializes the account ID of the caller into the scratch buffer.
```
[the remark this answers is not in the document]
Comment (pepyakin) [reply srml/contract/COMPLEXITY.md]: That's a good catch!

File: srml/contract/src/tests.rs [hunk 2]
```diff
@@ -869,3 +869,85 @@ fn input_data() {
...
+	(import "env" "ext_set_storage" (func $ext_set_storage (param i32 i32 i32 i32)))
+	(import "env" "memory" (memory 1 1))
+
+	(func (export "call")
+		;; Fill the scratch buffer with the caller.
+		(call $ext_caller)
+
+		;; Copy contents of the scratch buffer into the contract's memory.
+		(call $ext_scratch_copy
+			(i32.const 32)		;; Pointer in memory to the place where to copy.
```
[the remark this answers is not in the document]
Comment (pepyakin) [reply srml/contract/src/tests.rs]: the memory layout of that contract is the following:

[0..32]: the storage key (it is passed as the key for `ext_set_storage`)
[32..40]: contents of the scratch buffer (which is expected to be 8 bytes long)

But it is a good point, I'll add the comment about the memory layout and expand comments to the arguments.

File: srml/contract/src/vm/runtime.rs [hunk 3]
```diff
@@ -360,6 +360,16 @@ define_env!(init_env, <E: Ext>,
 		Err(sandbox::HostError)
 	},
 
+	// Stores the address of the caller into the scratch buffer.
+	//
+	// If this is a top-level call (i.e. initiated by an extrinsic) the origin address of the extrinsic
```
[the remark this answers is not in the document]
Comment (pepyakin) [reply srml/contract/src/vm/runtime.rs]: Do you have any use-cases for it in mind?

I'm aware of two use-cases for it:

1. authentification
2. figuring out whether caller is the origin (which basically says if it's human who calls you).

The first one as I've mentioned is a bad idea and I'm not sure if I'm convinced about the latter one.

[the remark this answers is not in the document]
Comment (Robbepop) [reply srml/contract/src/vm/runtime.rs]: Let me try to come up with some ideas and I will approach you then.

**Pull request merged**